src/Algebra/Graph/Acyclic/AdjacencyMap.hs

```diff
@@ -0,0 +1,418 @@
+-- Complexity: /O(L * log(L))/ time and /O(L)/ memory, where /L/ is
+-- the length of the given list.
+--
+-- @
+-- vertices []        == 'empty'
+-- vertices [x]       == 'vertex' x
+-- vertices [1, 2, 3] == 1 + 2 + 3
+-- @
+vertices :: (Ord a) => [a] -> AdjacencyMap a
+vertices = coerce AM.vertices
+
+-- | Perform a disjoint overlay of two different acyclic graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- edgeList (overlayD empty empty)             == []
+-- edgeList (overlayD (1 * 2 + 1 * 3) (1 * 2)) == [(Left 1,Left 2) 
+--                                                ,(Left 1,Left 3)
+--                                                ,(Right 1,Right 2)]
+-- @
+overlayD ::
+     (Ord a, Ord b)
+  => AdjacencyMap a
+  -> AdjacencyMap b
+  -> AdjacencyMap (Either a b)
+overlayD (AAM a) (AAM b) = AAM (AM.overlay (AM.gmap Left a) (AM.gmap Right b))
+
+-- | Perform a disjoint connect of two different acyclic graphs.
+-- Complexity: /O((n + m) * log(n))/ time and /O(n + m)/ memory.
+--
+-- @
+-- edgeList (connectD empty empty)     == []
+-- edgeList (connectD (1 + 2) (1 + 2)) == [(Left 1,Right 1) 
+--                                        ,(Left 1,Right 2)
+--                                        ,(Left 2,Right 1)
+--                                        ,(Left 2,Right 2)]
+-- edgeList (connectD (1 * 2) (1 * 2)) == [(Left 1,Left 2) 
+--                                        ,(Left 1,Right 1)
+--                                        ,(Left 1,Right 2) 
+--                                        ,(Left 2,Right 1) 
+--                                        ,(Left 2,Right 2) 
+--                                        ,(Right 1,Right 2)]
+-- @
+connectD ::
+     (Ord a, Ord b)
+  => AdjacencyMap a
+  -> AdjacencyMap b
+  -> AdjacencyMap (Either a b)
+connectD (AAM a) (AAM b) = AAM (AM.connect (AM.gmap Left a) (AM.gmap Right b))
+
+-- | Compute the /transitive closure/ of a graph.
+-- Complexity: /O(n * m * log(n)^2)/ time.
+--
+-- @
+-- transitiveClosure 'empty'               == 'empty'
+-- transitiveClosure ('vertex' x)          == 'vertex' x
+-- transitiveClosure (1 * 2 + 2 * 3)     == 1 * 2 + 2 * 3 + 1 * 3
+-- transitiveClosure . transitiveClosure == transitiveClosure
+-- @
+transitiveClosure :: Ord a => AdjacencyMap a -> AdjacencyMap a
+transitiveClosure = coerce AM.transitiveClosure
+
+-- | Compute the /condensation/ of a graph, where each vertex
+-- corresponds to a /strongly-connected component/ of the original
+-- graph. Note that component graphs are non-empty, and are therefore
+-- of type "Algebra.Graph.NonEmpty.AdjacencyMap".
+--
+-- @
+-- scc 'empty'      == 'empty'
+-- scc ('vertex' x) == 'vertex' (NonEmpty.'NonEmpty.vertex' x)
+-- @
+scc :: (Ord a) => AM.AdjacencyMap a -> AdjacencyMap (NonEmpty.AdjacencyMap a)
+scc = coerce AM.scc
+
+-- | Compute the /topological sort/ of a graph.
+-- 
+-- @
+-- topSort (1)             == [1]
+-- topSort (1 * 2 + 3 * 1) == [3,1,2]
```

Comment:
This seems to be a confusing example, because the `Num` instance is supposed to keep only the edge `(1, 2)`. Why do we get this order and not, say, `[1,2,3]`? By accident? The reader will surely be wondering about that.